File: Server/GameServer/MapCollision.cpp
#include "pch.h"
#include "MapCollision.h"

#include <algorithm>
#include <cmath>

MapCollision MapCollision::FromData(const MapCollisionData& data)
{
	MapCollision map;
	map._mapId = data.mapId;
	map._originX = data.originX;
	map._originY = data.originY;
	map._tileSize = data.tileSize > 0.f ? data.tileSize : 1.f;
	map._width = data.width;
	map._height = data.height;

	const size_t expectedSize = static_cast<size_t>(data.width) * static_cast<size_t>(data.height);
	map._blocked.resize(expectedSize, 0);

	for (size_t i = 0; i < expectedSize && i < data.cells.size(); ++i)
		map._blocked[i] = (data.cells[i] == '0') ? 1 : 0;

	return map;
}

bool MapCollision::IsInsideWorldBounds(float worldX, float worldY, float radius) const
{
	if (IsLoaded() == false)
		return false;

	const float minX = _originX + radius;
	const float maxX = _originX + static_cast<float>(_width) * _tileSize - radius;
	const float minY = _originY + radius;
	const float maxY = _originY + static_cast<float>(_height) * _tileSize - radius;

	if (minX > maxX || minY > maxY)
		return IsWalkable(worldX, worldY);

	return worldX >= minX && worldX <= maxX && worldY >= minY && worldY <= maxY;
}

bool MapCollision::IsTileWalkable(int32 tileX, int32 tileY) const
{
	if (tileX < 0 || tileY < 0 || tileX >= _width || tileY >= _height)
		return false;

	const size_t index = static_cast<size_t>(tileY) * static_cast<size_t>(_width) + static_cast<size_t>(tileX);
	if (index >= _blocked.size())
		return false;

	return _blocked[index] == 0;
}

void MapCollision::WorldToTile(float worldX, float worldY, int32& outTileX, int32& outTileY) const
{
	outTileX = static_cast<int32>(floorf((worldX - _originX) / _tileSize));
	outTileY = static_cast<int32>(floorf((worldY - _originY) / _tileSize));
}

bool MapCollision::IsWalkable(float worldX, float worldY) const
{
	if (IsLoaded() == false)
		return false;

	int32 tileX = 0;
	int32 tileY = 0;
	WorldToTile(worldX, worldY, tileX, tileY);
	return IsTileWalkable(tileX, tileY);
}

bool MapCollision::IsWalkableWithRadius(float worldX, float worldY, float radius) const
{
	if (IsLoaded() == false)
		return false;

	if (IsInsideWorldBounds(worldX, worldY, radius) == false)
		return false;

	if (radius <= 0.f)
		return IsWalkable(worldX, worldY);

	const float minX = worldX - radius;
	const float maxX = worldX + radius;
	const float minY = worldY - radius;
	const float maxY = worldY + radius;

	int32 minTileX = static_cast<int32>(floorf((minX - _originX) / _tileSize));
	int32 maxTileX = static_cast<int32>(floorf((maxX - _originX) / _tileSize));
	int32 minTileY = static_cast<int32>(floorf((minY - _originY) / _tileSize));
	int32 maxTileY = static_cast<int32>(floorf((maxY - _originY) / _tileSize));

	minTileX = std::max(minTileX, 0);
	minTileY = std::max(minTileY, 0);
	maxTileX = min(maxTileX, _width - 1);
	maxTileY = min(maxTileY, _height - 1);

	if (minTileX > maxTileX || minTileY > maxTileY)
		return false;

	for (int32 ty = minTileY; ty <= maxTileY; ++ty)
	{
		for (int32 tx = minTileX; tx <= maxTileX; ++tx)
		{
			if (IsTileWalkable(tx, ty) == false)
				return false;
		}
	}

	return true;
}

void MapCollision::ResolveMove(float fromX, float fromY, float toX, float toY, float radius, float& outX, float& outY) const
{
	if (IsLoaded() == false)
	{
		outX = fromX;
		outY = fromY;
		return;
	}

	if (IsWalkableWithRadius(toX, toY, radius))
	{
		outX = toX;
		outY = toY;
		return;
	}

	if (IsWalkableWithRadius(toX, fromY, radius))
	{
		outX = toX;
		outY = fromY;
		return;
	}

	if (IsWalkableWithRadius(fromX, toY, radius))
	{
		outX = fromX;
		outY = toY;
		return;
	}

	outX = fromX;
	outY = fromY;
}

File: Client/Assets/@Scripts/Data/MapCollision.cs
using UnityEngine;

public class MapCollision
{
    readonly int _mapId;
    readonly float _originX;
    readonly float _originY;
    readonly float _tileSize;
    readonly int _width;
    readonly int _height;
    readonly bool[] _blocked;

    public int MapId => _mapId;
    public bool IsLoaded => _width > 0 && _height > 0;

    public MapCollision(MapCollisionData data)
    {
        _mapId = data.mapId;
        _originX = data.originX;
        _originY = data.originY;
        _tileSize = data.tileSize > 0f ? data.tileSize : 1f;
        _width = data.width;
        _height = data.height;

        int expectedSize = data.width * data.height;
        _blocked = new bool[expectedSize];

        for (int i = 0; i < expectedSize && i < data.cells.Length; i++)
            _blocked[i] = data.cells[i] == Define.MAP_TOOL_WALL;
    }

    public bool IsWalkable(float worldX, float worldY)
    {
        if (IsLoaded == false)
            return false;

        WorldToTile(worldX, worldY, out int tileX, out int tileY);
        return IsTileWalkable(tileX, tileY);
    }

    public bool IsWalkableWithRadius(float worldX, float worldY, float radius)
    {
        if (IsLoaded == false)
            return false;

        if (IsInsideWorldBounds(worldX, worldY, radius) == false)
            return false;

        if (radius <= 0f)
            return IsWalkable(worldX, worldY);

        float minX = worldX - radius;
        float maxX = worldX + radius;
        float minY = worldY - radius;
        float maxY = worldY + radius;

        int minTileX = Mathf.FloorToInt((minX - _originX) / _tileSize);
        int maxTileX = Mathf.FloorToInt((maxX - _originX) / _tileSize);
        int minTileY = Mathf.FloorToInt((minY - _originY) / _tileSize);
        int maxTileY = Mathf.FloorToInt((maxY - _originY) / _tileSize);

        minTileX = Mathf.Max(minTileX, 0);
        minTileY = Mathf.Max(minTileY, 0);
        maxTileX = Mathf.Min(maxTileX, _width - 1);
        maxTileY = Mathf.Min(maxTileY, _height - 1);

        if (minTileX > maxTileX || minTileY > maxTileY)
            return false;

        for (int ty = minTileY; ty <= maxTileY; ty++)
        {
            for (int tx = minTileX; tx <= maxTileX; tx++)
            {
                if (IsTileWalkable(tx, ty) == false)
                    return false;
            }
        }

        return true;
    }

    public Vector2 ResolveMove(float fromX, float fromY, float toX, float toY, float radius)
    {
        if (IsLoaded == false)
            return new Vector2(fromX, fromY);

        if (IsWalkableWithRadius(toX, toY, radius))
            return new Vector2(toX, toY);

        if (IsWalkableWithRadius(toX, fromY, radius))
            return new Vector2(toX, fromY);

        if (IsWalkableWithRadius(fromX, toY, radius))
            return new Vector2(fromX, toY);

        return new Vector2(fromX, fromY);
    }

    public bool IsBlockedTile(int tileX, int tileY)
    {
        return IsTileWalkable(tileX, tileY) == false;
    }

    public float OriginX => _originX;
    public float OriginY => _originY;
    public float TileSize => _tileSize;
    public int Width => _width;
    public int Height => _height;

    bool IsInsideWorldBounds(float worldX, float worldY, float radius)
    {
        float minX = _originX + radius;
        float maxX = _originX + _width * _tileSize - radius;
        float minY = _originY + radius;
        float maxY = _originY + _height * _tileSize - radius;

        if (minX > maxX || minY > maxY)
            return IsWalkable(worldX, worldY);

        return worldX >= minX && worldX <= maxX && worldY >= minY && worldY <= maxY;
    }

    bool IsTileWalkable(int tileX, int tileY)
    {
        if (tileX < 0 || tileY < 0 || tileX >= _width || tileY >= _height)
            return false;

        int index = tileY * _width + tileX;
        if (index < 0 || index >= _blocked.Length)
            return false;

        return _blocked[index] == false;
    }

    void WorldToTile(float worldX, float worldY, out int tileX, out int tileY)
    {
        tileX = Mathf.FloorToInt((worldX - _originX) / _tileSize);
        tileY = Mathf.FloorToInt((worldY - _originY) / _tileSize);
    }
}
